tests/Kernel/MarkupRenderingTest.php

```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'components',
+    'filter',
+    'oe_bootstrap_theme_helper',
+    'system',
+    'ui_patterns',
+    'ui_patterns_library',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    // Replicate 'file_scan_ignore_directories' from settings.php.
+    $settings = Settings::getAll();
+    $settings['file_scan_ignore_directories'] = [
+      'node_modules',
+      'bower_components',
+      'vendor',
+      'build',
+      'bcl',
+    ];
+    new Settings($settings);
+
+    $this->container->get('theme_installer')->install(['oe_bootstrap_theme']);
+    $this->config('system.theme')->set('default', 'oe_bootstrap_theme')->save();
+    $this->container->set('theme.registry', NULL);
+  }
+
+  /**
+   * Tests rendering of elements.
+   *
+   * @param array $render_array
+   *   A render array.
+   * @param array $expectations
+   *   Test assertion expectations.
+   *
+   * @dataProvider markupRenderingProvider
+   */
+  public function testMarkupRendering(array $render_array, array $expectations): void {
+    // Wrap all the test structure inside a form. This will allow proper
+    // processing of form elements and invocation of form alter hooks. Even if
+    // the elements being tested are not form related, the form can host them
+    // without causing any issues.
+    $form_state = new FormState();
+    $form_state->addBuildInfo('args', [$render_array]);
+    $form_state->setProgrammed();
+
+    $form = $this->container->get('form_builder')->buildForm($this, $form_state);
+    $this->assertMarkupRendering($expectations, (string) $this->container->get('renderer')->renderRoot($form));
+  }
+
+  /**
+   * Data provider for markup rendering tests.
+   *
+   * @return array
+   *   A set of dump data for testing.
+   *
+   * @see self::testMarkupRendering()
+   * @see tests/fixtures/markup_rendering.yml
+   */
+  public function markupRenderingProvider(): array {
+    return Yaml::decode(file_get_contents(__DIR__ . '/../fixtures/markup_rendering.yml'));
+  }
+
+  /**
+   * Runs various assertion on given HTML string via CSS selectors.
+   *
+   * @param array $assertions
+   *   A HTML rendered markup.
+   * @param string $html
+   *   Test assertions. Associative array with having any of the following keys:
+   *   - count: assert how many times the given HTML elements occur.
+   *   - equals: assert content of given HTML elements.
+   *   - contains: assert content contained in given HTML elements.
+   *   Assertions array has to be provided in the following format:
+   *
+   *   @code
+   *   [
+   *     'count' => [
+   *       '.ecl-page-header' => 1,
+   *     ],
+   *     'equals' => [
+   *       '.ecl-page-header__identity' => 'Digital single market',
+   *     ],
+   *     'contains' => [
+   *       'Digital',
+   *       'single',
+   *       'market',
+   *     ],
+   *   ]
+   *   @endcode
```

Comment:
This sample contains `ecl` because it was copied from there. Let's adapt more to OE Bootstrap Theme.